minica/minica.go

```diff
@@ -0,0 +1,158 @@
+package minica
+
+import (
+	"crypto"
+	"crypto/x509"
+	"fmt"
+	"time"
+
+	"go.step.sm/crypto/sshutil"
+	"go.step.sm/crypto/x509util"
+	"golang.org/x/crypto/ssh"
+)
+
+// MiniCA is the implementation of a simple X.509 and SSH CA.
+type MiniCA struct {
+	Root          *x509.Certificate
+	Intermediate  *x509.Certificate
+	Signer        crypto.Signer
+	SSHHostSigner ssh.Signer
+	SSHUserSigner ssh.Signer
+}
+
+// New creates a new MiniCA, the custom options allows to overwrite templates,
+// signer types and certificate names.
+func New(opts ...Option) (*MiniCA, error) {
+	now := time.Now()
+	o := newOptions().apply(opts)
+
+	// Create root
+	rootSubject := o.Name + " Root CA"
+	rootSigner, err := o.GetSigner()
+	if err != nil {
+		return nil, err
+	}
+	rootCR, err := x509util.CreateCertificateRequest(rootSubject, []string{}, rootSigner)
+	if err != nil {
+		return nil, err
+	}
+	cert, err := x509util.NewCertificate(rootCR, x509util.WithTemplate(o.RootTemplate, x509util.CreateTemplateData(rootSubject, []string{})))
+	if err != nil {
+		return nil, err
+	}
+	template := cert.GetCertificate()
+	template.NotBefore = now
+	template.NotAfter = now.Add(24 * time.Hour)
+	root, err := x509util.CreateCertificate(template, template, rootSigner.Public(), rootSigner)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create intermediate
+	intSubject := o.Name + " Intermediate CA"
+	intSigner, err := o.GetSigner()
+	if err != nil {
+		return nil, err
+	}
+	intCR, err := x509util.CreateCertificateRequest(intSubject, []string{}, intSigner)
+	if err != nil {
+		return nil, err
+	}
+	cert, err = x509util.NewCertificate(intCR, x509util.WithTemplate(o.IntermediateTemplate, x509util.CreateTemplateData(intSubject, []string{})))
+	if err != nil {
+		return nil, err
+	}
+	template = cert.GetCertificate()
+	template.NotBefore = now
+	template.NotAfter = now.Add(24 * time.Hour)
+	intermediate, err := x509util.CreateCertificate(template, root, intSigner.Public(), rootSigner)
+	if err != nil {
+		return nil, err
+	}
+
+	// Ssh host signer
+	signer, err := o.GetSigner()
+	if err != nil {
+		return nil, err
+	}
+	sshHostSigner, err := ssh.NewSignerFromSigner(signer)
+	if err != nil {
+		return nil, err
+	}
+
+	// Ssh user signer
+	signer, err = o.GetSigner()
+	if err != nil {
+		return nil, err
+	}
+	sshUserSigner, err := ssh.NewSignerFromSigner(signer)
+	if err != nil {
+		return nil, err
+	}
+
+	return &MiniCA{
+		Root:          root,
+		Intermediate:  intermediate,
+		Signer:        intSigner,
+		SSHHostSigner: sshHostSigner,
+		SSHUserSigner: sshUserSigner,
+	}, nil
+}
+
+// Sign signs an X.509 certificate template using the intermediate certificate.
+func (c *MiniCA) Sign(template *x509.Certificate) (*x509.Certificate, error) {
+	if template.NotBefore.IsZero() {
+		template.NotBefore = time.Now()
```

Comment:
I think since we may end up mutating the input we should document the function to include this case?

Reply:
Fixed with https://github.com/smallstep/crypto/commit/cd2e52da0b66ce3a0fde8470ba7751d1e9a970eb